src/Microsoft.AspNetCore.Razor.Evolution/ITagHelperBoundAttributeDescriptorBuilder.cs

```diff
@@ -0,0 +1,282 @@
+            [typeof(decimal).FullName] = "decimal",
+        };
+
+        private static ICollection<char> InvalidNonWhitespaceAttributeNameCharacters { get; } = new HashSet<char>(
+            new[] { '@', '!', '<', '/', '?', '[', '>', ']', '=', '"', '\'', '*' });
+
+        private bool _isEnum;
+        private string _indexerValueTypeName;
+        private string _name;
+        private string _propertyName;
+        private string _typeName;
+        private string _documentation;
+        private string _indexerNamePrefix;
+        private readonly string _containingTypeName;
+        private readonly Dictionary<string, string> _metadata;
+        private HashSet<RazorDiagnostic> _diagnostics;
+
+        private ITagHelperBoundAttributeDescriptorBuilder(string containingTypeName)
+        {
+            _containingTypeName = containingTypeName;
+            _metadata = new Dictionary<string, string>();
+        }
+
+        public static ITagHelperBoundAttributeDescriptorBuilder Create(string containingTypeName)
+        {
+            return new ITagHelperBoundAttributeDescriptorBuilder(containingTypeName);
+        }
+
+        public ITagHelperBoundAttributeDescriptorBuilder Name(string name)
+        {
+            _name = name;
+
+            return this;
+        }
+
+        public ITagHelperBoundAttributeDescriptorBuilder PropertyName(string propertyName)
+        {
+            _propertyName = propertyName;
+
+            return this;
+        }
+
+        public ITagHelperBoundAttributeDescriptorBuilder TypeName(string typeName)
+        {
+            _typeName = typeName;
+
+            return this;
+        }
+
+        public ITagHelperBoundAttributeDescriptorBuilder AsEnum()
+        {
+            _isEnum = true;
+
+            return this;
+        }
+
+        public ITagHelperBoundAttributeDescriptorBuilder AsDictionary(string attributeNamePrefix, string valueTypeName)
+        {
+            _indexerNamePrefix = attributeNamePrefix;
+            _indexerValueTypeName = valueTypeName;
+
+            return this;
+        }
+
+        public ITagHelperBoundAttributeDescriptorBuilder Documentation(string documentation)
+        {
+            _documentation = documentation;
+
+            return this;
+        }
+
+        public ITagHelperBoundAttributeDescriptorBuilder AddMetadata(string key, string value)
+        {
+            _metadata[key] = value;
+
+            return this;
+        }
+
+        public ITagHelperBoundAttributeDescriptorBuilder AddDiagnostic(RazorDiagnostic diagnostic)
+        {
+            EnsureDiagnostics();
+            _diagnostics.Add(diagnostic);
+
+            return this;
+        }
+
+        public BoundAttributeDescriptor Build()
+        {
+            var validationDiagnostics = Validate();
+
+            var diagnostics = new List<RazorDiagnostic>();
+            diagnostics.AddRange(validationDiagnostics);
+
+            if (_diagnostics != null)
```

Comment:
Could just iterate over each of the validationDiagnostics and call `AddDiagnostic` which will ensure `_diagnostics` is setup appropriately; will be more clear. Same in other builders.